gateway/ln-gateway/src/lightning/ldk.rs

```diff
@@ -0,0 +1,501 @@
+        if let Err(err) = self.node.stop() {
+            error!("Failed to stop LDK Node: {err:?}");
+        } else {
+            info!("LDK Node stopped.");
+        }
+    }
+}
+
+#[async_trait]
+impl ILnRpcClient for GatewayLdkClient {
+    async fn info(&self) -> Result<GetNodeInfoResponse, LightningRpcError> {
+        let node_status = self.node.status();
+
+        let chain_tip_block_summary = self
+            .esplora_client
+            .get_blocks(None)
+            .await
+            .unwrap()
+            .into_iter()
+            .next()
+            .unwrap();
+
+        let esplora_chain_tip_timestamp = chain_tip_block_summary.time.timestamp;
+        let block_height: u32 = chain_tip_block_summary.time.height;
+
+        let synced_to_chain = node_status.latest_wallet_sync_timestamp.unwrap_or_default()
+            > esplora_chain_tip_timestamp
+            && node_status
+                .latest_onchain_wallet_sync_timestamp
+                .unwrap_or_default()
+                > esplora_chain_tip_timestamp;
+
+        Ok(GetNodeInfoResponse {
+            pub_key: self.node.node_id().serialize().to_vec(),
+            // TODO: This is a placeholder. We need to get the actual alias from the LDK node.
+            alias: "LDK Fedimint Gateway Node".to_string(),
+            network: match self.node.config().network {
+                Network::Bitcoin => "main",
+                Network::Testnet => "test",
+                Network::Signet => "signet",
+                Network::Regtest => "regtest",
+                _ => panic!("Unsupported network"),
+            }
+            .to_string(),
+            block_height,
+            synced_to_chain,
+        })
+    }
+
+    async fn routehints(
+        &self,
+        _num_route_hints: usize,
+    ) -> Result<GetRouteHintsResponse, LightningRpcError> {
+        Ok(GetRouteHintsResponse {
+            route_hints: vec![],
+        })
+    }
+
+    async fn pay(
+        &self,
+        invoice: PayInvoiceRequest,
+    ) -> Result<PayInvoiceResponse, LightningRpcError> {
+        let parsed_invoice: Bolt11Invoice =
+            Bolt11Invoice::from_str(&invoice.invoice).map_err(|e| {
+                LightningRpcError::FailedPayment {
+                    failure_reason: format!("Failed to parse invoice: {e:?}"),
+                }
+            })?;
+
+        let payment_id = match self.node.bolt11_payment().send(&parsed_invoice) {
+            Ok(payment_id) => payment_id,
+            Err(e) => {
+                return Err(LightningRpcError::FailedPayment {
+                    failure_reason: format!("LDK payment failed to initialize: {e:?}"),
+                });
+            }
+        };
+
+        // TODO: Find a way to avoid looping/polling to find out when a payment is
+        // completed. `ldk-node` provides `PaymentSuccessful` and `PaymentFailed`
+        // events, but routing them to specific payment IDs isn't straightforward.
```

Comment:
What does "routing them" mean here?

Reply:
Sorry, that was a strange wording. What I mean is we should only be handling events from `ldk-node`'s event queue in one place to avoid having two threads stomping/skipping over events. Right now, event handling is done in `GatewayLdkClient::event_handler_task_handle`, so we'd need to add functionality there to handle `PaymentSuccessful` and `PaymentFailed` events and somehow be able to tell `GatewayLdkClient::pay` when a payment has succeeded or failed. This can definitely be done, but we'd need some sort of per-paymenthash queue or something.

I cleaned up the comment here